In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('tradelog.csv',index_col=[0] )

In [3]:
df

Ticker        Entry Time  Entry Price         Exit Time  \
1    BANKNIFTY09JUL2022800PE  08-07-2020 09:29       163.50  08-07-2020 13:15   
2    BANKNIFTY09JUL2022700CE  09-07-2020 09:29       162.50  09-07-2020 10:34   
3    BANKNIFTY09JUL2022700PE  09-07-2020 09:29       110.55  09-07-2020 10:34   
4    BANKNIFTY16JUL2022700CE  10-07-2020 09:29       295.10  10-07-2020 15:24   
5    BANKNIFTY16JUL2022700PE  10-07-2020 09:29       555.20  10-07-2020 15:24   
..                       ...               ...          ...               ...   
245  BANKNIFTY31DEC2031300PE  29-12-2020 09:29       216.65  29-12-2020 13:10   
246  BANKNIFTY31DEC2031200CE  30-12-2020 09:29       248.25  30-12-2020 10:24   
247  BANKNIFTY31DEC2031200PE  30-12-2020 09:29       221.70  30-12-2020 10:24   
248  BANKNIFTY31DEC2031300CE  31-12-2020 09:29       161.80  31-12-2020 10:37   
249  BANKNIFTY31DEC2031300PE  31-12-2020 09:29       105.05  31-12-2020 10:37   

     Exit Price  
1        301.00  
2        166.75  
3        147.00  
4        437.55  
5        403.15  
..          ...  
245      274.60  
246      283.90  
247      240.10  
248      173.85  
249      124.00  

[249 rows x 5 columns]

In [4]:
df.isnull().sum()

Ticker         0
Entry Time     0
Entry Price    0
Exit Time      0
Exit Price     0
dtype: int64

In [5]:
df['Entry Time'] = pd.to_datetime(df['Entry Time'])
df['Exit Time'] = pd.to_datetime(df['Exit Time'])

In [6]:
df = df.sort_values(by='Entry Time')
df = df.reset_index(drop=True)
df

Ticker          Entry Time  Entry Price  \
0    BANKNIFTY03SEP2024100CE 2020-01-09 09:29:00       288.50   
1    BANKNIFTY03SEP2024100PE 2020-01-09 09:29:00       703.80   
2    BANKNIFTY01OCT2021900CE 2020-01-10 09:29:00       110.35   
3    BANKNIFTY01OCT2021900PE 2020-01-10 09:29:00        87.05   
4    BANKNIFTY03DEC2029600PE 2020-01-12 09:29:00       284.10   
..                       ...                 ...          ...   
244  BANKNIFTY31DEC2031300PE 2020-12-29 09:29:00       216.65   
245  BANKNIFTY31DEC2031200CE 2020-12-30 09:29:00       248.25   
246  BANKNIFTY31DEC2031200PE 2020-12-30 09:29:00       221.70   
247  BANKNIFTY31DEC2031300CE 2020-12-31 09:29:00       161.80   
248  BANKNIFTY31DEC2031300PE 2020-12-31 09:29:00       105.05   

              Exit Time  Exit Price  
0   2020-01-09 09:57:00      440.70  
1   2020-01-09 09:57:00      450.10  
2   2020-01-10 10:01:00      114.55  
3   2020-01-10 10:01:00      118.65  
4   2020-01-12 13:08:00      428.75  
..                  ...         ...  
244 2020-12-29 13:10:00      274.60  
245 2020-12-30 10:24:00      283.90  
246 2020-12-30 10:24:00      240.10  
247 2020-12-31 10:37:00      173.85  
248 2020-12-31 10:37:00      124.00  

[249 rows x 5 columns]

In [7]:
df['Total_Price'] = df['Exit Price'] - df['Entry Price']

In [8]:
price_list = list(df['Total_Price'])
price_list

[152.2,
 -253.69999999999993,
 4.200000000000003,
 31.60000000000001,
 144.64999999999998,
 -50.75,
 73.0,
 -12.0,
 158.54999999999995,
 -254.89999999999998,
 46.099999999999966,
 18.05000000000001,
 -82.79999999999995,
 100.05000000000001,
 43.95,
 -17.19999999999999,
 25.30000000000001,
 71.39999999999998,
 -51.900000000000006,
 86.70000000000002,
 -158.60000000000002,
 198.24999999999997,
 -72.55000000000001,
 170.39999999999998,
 22.80000000000001,
 60.35000000000002,
 17.19999999999999,
 76.39999999999998,
 79.15,
 -132.25,
 38.599999999999966,
 35.150000000000034,
 5.400000000000006,
 28.80000000000001,
 -70.35000000000002,
 33.55000000000001,
 -9.349999999999994,
 67.89999999999998,
 -212.25,
 122.45000000000005,
 -111.80000000000001,
 172.74999999999997,
 26.25,
 50.44999999999999,
 -186.75000000000003,
 115.60000000000001,
 -89.85000000000002,
 135.25000000000003,
 -234.35000000000002,
 141.20000000000002,
 90.29999999999998,
 -185.09999999999997,
 125.0,
 -63.60000000000002,


In [9]:
sum2 = 6500
list1 = []
for i in price_list:
    sum2 = sum2 + i
    list1.append(sum2)
df['port_final'] = list1

In [10]:
df

Ticker          Entry Time  Entry Price  \
0    BANKNIFTY03SEP2024100CE 2020-01-09 09:29:00       288.50   
1    BANKNIFTY03SEP2024100PE 2020-01-09 09:29:00       703.80   
2    BANKNIFTY01OCT2021900CE 2020-01-10 09:29:00       110.35   
3    BANKNIFTY01OCT2021900PE 2020-01-10 09:29:00        87.05   
4    BANKNIFTY03DEC2029600PE 2020-01-12 09:29:00       284.10   
..                       ...                 ...          ...   
244  BANKNIFTY31DEC2031300PE 2020-12-29 09:29:00       216.65   
245  BANKNIFTY31DEC2031200CE 2020-12-30 09:29:00       248.25   
246  BANKNIFTY31DEC2031200PE 2020-12-30 09:29:00       221.70   
247  BANKNIFTY31DEC2031300CE 2020-12-31 09:29:00       161.80   
248  BANKNIFTY31DEC2031300PE 2020-12-31 09:29:00       105.05   

              Exit Time  Exit Price  Total_Price  port_final  
0   2020-01-09 09:57:00      440.70       152.20     6652.20  
1   2020-01-09 09:57:00      450.10      -253.70     6398.50  
2   2020-01-10 10:01:00      114.55         4.20     6402.70  
3   2020-01-10 10:01:00      118.65        31.60     6434.30  
4   2020-01-12 13:08:00      428.75       144.65     6578.95  
..                  ...         ...          ...         ...  
244 2020-12-29 13:10:00      274.60        57.95     7925.35  
245 2020-12-30 10:24:00      283.90        35.65     7961.00  
246 2020-12-30 10:24:00      240.10        18.40     7979.40  
247 2020-12-31 10:37:00      173.85        12.05     7991.45  
248 2020-12-31 10:37:00      124.00        18.95     8010.40  

[249 rows x 7 columns]

## 1. Total Trades

In [11]:
total_t = df.shape[0]
total_t

249

## 2. Profitable Trades

In [12]:
prof_df = df[df['Total_Price']>0]
prof_df = prof_df.reset_index(drop = True)
prof_df

Ticker          Entry Time  Entry Price  \
0    BANKNIFTY03SEP2024100CE 2020-01-09 09:29:00       288.50   
1    BANKNIFTY01OCT2021900CE 2020-01-10 09:29:00       110.35   
2    BANKNIFTY01OCT2021900PE 2020-01-10 09:29:00        87.05   
3    BANKNIFTY03DEC2029600PE 2020-01-12 09:29:00       284.10   
4    BANKNIFTY03SEP2023700CE 2020-02-09 09:29:00       238.00   
..                       ...                 ...          ...   
148  BANKNIFTY31DEC2031300PE 2020-12-29 09:29:00       216.65   
149  BANKNIFTY31DEC2031200CE 2020-12-30 09:29:00       248.25   
150  BANKNIFTY31DEC2031200PE 2020-12-30 09:29:00       221.70   
151  BANKNIFTY31DEC2031300CE 2020-12-31 09:29:00       161.80   
152  BANKNIFTY31DEC2031300PE 2020-12-31 09:29:00       105.05   

              Exit Time  Exit Price  Total_Price  port_final  
0   2020-01-09 09:57:00      440.70       152.20     6652.20  
1   2020-01-10 10:01:00      114.55         4.20     6402.70  
2   2020-01-10 10:01:00      118.65        31.60     6434.30  
3   2020-01-12 13:08:00      428.75       144.65     6578.95  
4   2020-02-09 10:46:00      311.00        73.00     6601.20  
..                  ...         ...          ...         ...  
148 2020-12-29 13:10:00      274.60        57.95     7925.35  
149 2020-12-30 10:24:00      283.90        35.65     7961.00  
150 2020-12-30 10:24:00      240.10        18.40     7979.40  
151 2020-12-31 10:37:00      173.85        12.05     7991.45  
152 2020-12-31 10:37:00      124.00        18.95     8010.40  

[153 rows x 7 columns]

In [13]:
profit = prof_df.shape[0]
profit

153

### There are 153 Profit Trades

## 3. Loss Trades

In [14]:
loss_df = df[df['Total_Price']<0]

In [15]:
loss_df = loss_df.reset_index(drop = True)
loss_df

Ticker          Entry Time  Entry Price  \
0   BANKNIFTY03SEP2024100PE 2020-01-09 09:29:00       703.80   
1   BANKNIFTY03DEC2029600CE 2020-01-12 09:29:00       459.20   
2   BANKNIFTY03SEP2023700PE 2020-02-09 09:29:00       302.25   
3   BANKNIFTY05NOV2024200CE 2020-02-11 09:29:00       697.55   
4   BANKNIFTY06AUG2021400PE 2020-03-08 09:29:00       463.15   
..                      ...                 ...          ...   
91  BANKNIFTY24DEC2030300PE 2020-12-21 09:29:00       498.35   
92  BANKNIFTY24DEC2029600CE 2020-12-23 09:29:00       306.85   
93  BANKNIFTY24DEC2030200CE 2020-12-24 09:29:00       254.10   
94  BANKNIFTY31DEC2030800CE 2020-12-28 09:29:00       348.15   
95  BANKNIFTY31DEC2031300CE 2020-12-29 09:29:00       274.75   

             Exit Time  Exit Price  Total_Price  port_final  
0  2020-01-09 09:57:00      450.10      -253.70     6398.50  
1  2020-01-12 13:08:00      408.45       -50.75     6528.20  
2  2020-02-09 10:46:00      290.25       -12.00     6589.20  
3  2020-02-11 11:19:00      442.65      -254.90     6492.85  
4  2020-03-08 15:24:00      380.35       -82.80     6474.20  
..                 ...         ...          ...         ...  
91 2020-12-21 13:37:00      309.05      -189.30     7702.95  
92 2020-12-23 14:13:00      218.15       -88.70     7697.10  
93 2020-12-24 10:15:00      139.45      -114.65     7801.80  
94 2020-12-28 13:43:00      328.80       -19.35     7782.45  
95 2020-12-29 13:10:00      272.15        -2.60     7867.40  

[96 rows x 7 columns]

In [16]:
loss = loss_df.shape[0]
loss

96

### There are 96 Loss-Making Trades

## 4 Win rate: Win Rate is the ratio of profitable trades to all trades

In [17]:
winRate = prof_df.shape[0]/df.shape[0] 
print(winRate)

0.6144578313253012


### Winrate = 0.6144

## 5 Average Profit per trade

In [18]:
sum1 = prof_df['Total_Price'].sum()
avgProf = sum1/prof_df.shape[0]
avgProf

76.09542483660131

### Average Profit per trade is 76.0954

## 6. Average Loss per trade

In [19]:
sum2 = loss_df['Total_Price'].sum()
avgloss = sum2/loss_df.shape[0]
avgloss

-105.54375

### Average Loss per trade is -105.54375

## 7. Risk Reward ratio: Risk Reward Ratio is the ratio of average profit to average loss.

In [20]:
RiskRew = avgProf/avgloss
RiskRew

-0.720984661210174

### 8. Expectancy : (Win Rate x Average Profit) - (Loss Rate x Average Loss), where Loss Rate = 1 - (Win Rate)

In [21]:
expec = (winRate * avgProf) - ((1-winRate) * avgloss)
expec

87.44899598393575

## 9. Average ROR per trade:

In [22]:
import statistics

volatality = statistics.stdev(df['Total_Price'])
volatality

113.06325065442704

In [23]:
AvgReturn = df['Total_Price'].sum() / df.shape[0]
AvgReturn

6.06586345381526

In [24]:
riskreturn = 0.05
ROR = (AvgReturn - riskreturn) /volatality
ROR

0.05320794704729027

## 10. Sharpe Ratio:

In [25]:
total_return = df['Total_Price'].sum()
total_return

1510.3999999999999

In [26]:
Ret_init = 6500 #given
Ret_fin = Ret_init + total_return
Ret_fin

8010.4

In [27]:
rateofret = ((Ret_fin - Ret_init)/Ret_init )
rateofret

0.23236923076923072

In [28]:
sharpe = (rateofret - riskreturn)/volatality
sharpe

0.0016129841457206502

## 11. Max Drawdown

In [29]:
maxi = df['port_final'].max() - df['port_final'].min()
maxi

1618.4499999999962

## 12. Max Drawdown Percentage

In [30]:
# df['port_final'] = df['Total_Price'] + Ret_init #6500
df

Ticker          Entry Time  Entry Price  \
0    BANKNIFTY03SEP2024100CE 2020-01-09 09:29:00       288.50   
1    BANKNIFTY03SEP2024100PE 2020-01-09 09:29:00       703.80   
2    BANKNIFTY01OCT2021900CE 2020-01-10 09:29:00       110.35   
3    BANKNIFTY01OCT2021900PE 2020-01-10 09:29:00        87.05   
4    BANKNIFTY03DEC2029600PE 2020-01-12 09:29:00       284.10   
..                       ...                 ...          ...   
244  BANKNIFTY31DEC2031300PE 2020-12-29 09:29:00       216.65   
245  BANKNIFTY31DEC2031200CE 2020-12-30 09:29:00       248.25   
246  BANKNIFTY31DEC2031200PE 2020-12-30 09:29:00       221.70   
247  BANKNIFTY31DEC2031300CE 2020-12-31 09:29:00       161.80   
248  BANKNIFTY31DEC2031300PE 2020-12-31 09:29:00       105.05   

              Exit Time  Exit Price  Total_Price  port_final  
0   2020-01-09 09:57:00      440.70       152.20     6652.20  
1   2020-01-09 09:57:00      450.10      -253.70     6398.50  
2   2020-01-10 10:01:00      114.55         4.20     6402.70  
3   2020-01-10 10:01:00      118.65        31.60     6434.30  
4   2020-01-12 13:08:00      428.75       144.65     6578.95  
..                  ...         ...          ...         ...  
244 2020-12-29 13:10:00      274.60        57.95     7925.35  
245 2020-12-30 10:24:00      283.90        35.65     7961.00  
246 2020-12-30 10:24:00      240.10        18.40     7979.40  
247 2020-12-31 10:37:00      173.85        12.05     7991.45  
248 2020-12-31 10:37:00      124.00        18.95     8010.40  

[249 rows x 7 columns]

In [31]:
draw_max = maxi
peak = max(df['port_final'])

print(draw_max, peak)

1618.4499999999962 8016.949999999996


In [32]:
m_draw_per = (draw_max / peak) * 100
m_draw_per

20.187851988599117

## 13. CAGR

In [33]:
df['Exit Time'].max()

Timestamp('2020-12-31 10:37:00')

In [34]:
df['Exit Time'].min()

Timestamp('2020-01-09 09:57:00')

In [35]:
days = (df['Exit Time'].max() - df['Exit Time'].min()).days

In [36]:
days

357

In [37]:
ending = df.iloc[248,6]
ending

8010.399999999994

In [38]:
initial = 6500


In [39]:
cagr = ((ending / initial) ** (1/days)) - 1
cagr

0.0005854332614114721

## 14. Calmar Ratio:

 Calmar ratio = ( Compound Annual Return) / (Maximum drawdown)
 
 CAR = [(Ending Value / Beginning Value) ^ (1 / Number of Years)] - 1

In [40]:
car = ((ending / 6500) ** (1/1)) - 1 # for year 2020
car

0.23236923076922977

In [41]:
calmer = car / maxi
calmer

0.0001435751680739166

In [42]:

new_dict = {
    'Total Trades' : total_t,
    'Profitable Trades' : profit,
    'Loss-Making Trades' : loss,
    'Win rate' : winRate,
    'Average Profit per trade' : avgProf,
    'Average Loss per trade' : avgloss,
    'Risk Reward ratio' : RiskRew,
    'Expectancy' : expec,
    'Average ROR per trade' : ROR,
    'Sharpe Ratio' : sharpe,
    'Max Drawdown' : maxi,
    'Max Drawdown Percentage' : m_draw_per,
    'CAGR' : cagr,
    'Calmar Ratio' : calmer
}

final_df = pd.DataFrame(list(new_dict.keys()), columns=['trade'])
final_df['values'] = new_dict.values()
final_df

trade       values
0               Total Trades   249.000000
1          Profitable Trades   153.000000
2         Loss-Making Trades    96.000000
3                   Win rate     0.614458
4   Average Profit per trade    76.095425
5     Average Loss per trade  -105.543750
6          Risk Reward ratio    -0.720985
7                 Expectancy    87.448996
8      Average ROR per trade     0.053208
9               Sharpe Ratio     0.001613
10              Max Drawdown  1618.450000
11   Max Drawdown Percentage    20.187852
12                      CAGR     0.000585
13              Calmar Ratio     0.000144

In [43]:
final_df.to_csv('Task1_Aman.csv',index=False)